### ISSUE TITLE
AsyncGRPO has no importance-sampling correction, but TRL still imposes --logprobs-mode processed_logprobs on its server

### ISSUE THREAD

**Issue by behroozazarkhalili** [huggingface/trl#6945]
### The setting

TRL imposes `--logprobs-mode processed_logprobs` on every vLLM server it talks to. `trl.scripts.vllm_serve.build_command` appends it unconditionally, and `vllm_serve.py:201` lists it as one of "three settings imposed by TRL", with the reason given as:

> so that temperature scaling and logit tweaking are reflected in the returned logprobs, which trainers use for importance sampling correction

`docs/source/vllm_integration.md:183` repeats it: "which is what the importance sampling correction expects."

### The problem

That rationale is specific to the importance-sampling correction, and `AsyncGRPOTrainer` does not have one. `grep importance_sampling` returns nothing in either `async_grpo_config.py` or `async_grpo_trainer.py`.

What AsyncGRPO does instead is use vLLM's logprob directly as the PPO denominator:

```python
# trl/experimental/async_grpo/async_grpo_trainer.py:1113
log_ratio = log_probs - old_log_probs
coef_1 = torch.exp(log_ratio)
coef_2 = torch.clamp(coef_1, 1 - self.epsilon_low, 1 + self.epsilon_high)
```

`old_log_probs` is collated straight from the rollout logprobs, and `log_probs` is a full-vocab log-softmax. With processed logprobs the two normalize over different supports, so for an unchanged policy the ratio reads the surviving mass `S` rather than `1.0`:

| `top_p` | `coef_1` | error vs 1.0 | outside `(0.8, 1.2)`? |
| --- | --- | --- | --- |
| 1.00 | 1.0000 | 0.0000 | no |
| 0.90 | 0.9000 | 0.1000 | no |
| 0.80 | 0.8000 | 0.2000 | no |
| 0.70 | 0.7000 | 0.3000 | yes |

Below `1 - epsilon_low` the clip engages on every negative-advantage token whether or not the policy moved.

The contrast with `GRPOTrainer` is what makes this a design question rather than a bug report. GRPO computes its own `old_per_token_logps` (`grpo_trainer.py:2664`) and differences it against vLLM's (`:2680`), so processed logprobs are exactly what it wants and `vllm_importance_sampling_correction=False` disables the whole path. AsyncGRPO gets the setting imposed on it and has no switch.

### Proposal

Make the imposed flag conditional on the trainer having an importance-sampling correction, rather than applying it to every server TRL prescribes. Trainers that difference two logprob sources want `processed_logprobs`; trainers that consume the logprob as a behavior-policy density want the raw one.

I have not been able to schedule a two-GPU async run to measure this end to end, so the table above is derived from the two normalizations and computed exactly rather than observed. Happy to run it if someone can point me at a lane.

#6944 adds a warning on the AsyncGRPO side in the meantime. It does not change the flag, since that is this decision to make.

**Comment by verma8076** [comment]
Checked whether "conditional on the trainer having a correction" has a clean target to switch to, from vLLM's sampler source (`vllm/v1/sample/sampler.py`, `vllm/v1/sample/ops/topk_topp_sampler.py`).

There isn't one. `raw_logprobs` isn't "pre-truncation, post-temperature" — it's computed before *any* processing at all:

```python
# NOTE(woosuk): Use the original logits (before any penalties or
# temperature scaling) for the top-k logprobs.
raw_logprobs: torch.Tensor | None = None
if num_logprobs is not None or sampling_metadata.logprob_token_ids:
    if logprobs_mode == "raw_logprobs":
        raw_logprobs = self.compute_logprobs(logits)
```

That branch runs *before* `apply_logits_processors`, penalties, and `apply_temperature` — all of which happen later inside `sample()`. `processed_logprobs`, by contrast, is computed inside `topk_topp_sampler` after temperature, the argmax-invariant processors (min_p), and top_k/top_p have all been applied. So the two modes bracket the whole pipeline; there's no vLLM mode that gives temperature-scaled-but-not-truncated logprobs.

If AsyncGRPO switched its server to `raw_logprobs` to dodge the truncation bias, `old_log_probs` would then come from the un-scaled distribution while `log_probs` (the trainer's own recompute) divides by temperature before its log-softmax — the same shape of mismatch this issue describes, just from temperature instead of top_p/top_k/min_p/repetition_penalty, and larger for any temperature meaningfully off 1.0 (temperature reshapes the full distribution, not just a tail). `AsyncGRPOConfig.temperature` defaults to `1.0` but that's not what real RL rollouts run at.

So "conditional flag" as stated trades one bias for another rather than removing it. Making it actually work would need either a fifth vLLM logprobs mode (temperature-only), or AsyncGRPO applying temperature itself to raw logprobs before differencing (recoverable, since `T` is a known scalar and `raw_logits / T` then log-softmax reproduces the temperature-scaled distribution without the truncation renormalization) — the second one seems achievable without touching vLLM at all. Haven't verified that end-to-end either, same caveat as the table above.

**Comment by mmjerge** [comment]
I'd like to implement the correction, and I've studied both loss paths closely enough to see that the sync scheme cannot simply be transplanted. Here is a concrete design; I'd welcome maintainer direction on the one genuine fork in the road before I open a PR.

**Why AsyncGRPO can't mirror sync GRPO directly.** The synchronous trainer keeps two ratios apart: a PPO drift ratio built from *recomputed* old logprobs (`grpo_trainer.py:2649-2665`), and a separate trainer-vs-sampler correction `exp(old_per_token_logps − sampling_per_token_logps)` that multiplies the loss (`:3226-3227`). The recomputation is the load-bearing part: vLLM's logprobs are never trusted as π_old, only as π_sampler. AsyncGRPO cannot do this. By the time a sample leaves the rollout queue, the weights that generated it are up to `max_staleness` versions gone, so there is nothing to recompute π_old(θ_gen) with — short of snapshotting every intermediate model version, which nobody wants to pay for. The combined ratio π_θ/π_vLLM(θ_stale) at `async_grpo_trainer.py:1140` is not an oversight; it is the only ratio available.

**What the combined ratio costs us.** Two things. First, the top-p support mismatch this issue documents: an unchanged policy reads ratio ≈ S, so with `processed_logprobs` and top_p = 0.9 every token starts 10% off before any drift. Second, and worse: for negative advantages the PPO min-clip leaves the ratio unbounded above — `-min(rA, clip(r)A)` with A < 0 grows linearly in r. Staleness times sampler mismatch is exactly the regime that produces large r, so the trainer's variance hazard concentrates on the tokens it most needs to damp.

**Proposal.** Apply the existing TIS/MIS family to the combined ratio as a detached multiplicative gate on the per-token loss, configured through the same four fields sync GRPO already exposes (`vllm_importance_sampling_correction`, `_mode`, `_clip_max`, `_clip_min`), added to `AsyncGRPOConfig` with matching names and validation. Mechanics:

- Build ρ = `coef_1.detach()` after `async_grpo_trainer.py:1141`; for the `sequence_*` modes, sum log-ratios per sequence using `position_ids` segment boundaries (the padding-free packing means a segment scatter, not a `(B, 1)` broadcast).
- `*_mask` modes multiply the loss by the indicator `C_min ≤ ρ ≤ C_max` (ICEPOP-style); `*_truncate` modes multiply by `clamp(ρ, C_min, C_max) / ρ`, which caps the effective gradient weight at the bound while leaving in-range tokens untouched.
- Log the same diagnostics as sync (ratio min/mean/max, masked fraction), which also gives users their first visibility into how far off-policy their queue actually runs.
- This resolves the `processed_logprobs` question in favor of keeping the imposed flag: with a correction in place, the flag's stated rationale becomes true for AsyncGRPO rather than vacuous.

**The fork in the road.** The gate treats the *same* number as both the PPO ratio and the mismatch estimate. That conflation is inherent to async (see above), and it is what the masked-IS literature prescribes for exactly this setting, but it means `_mask` here gates on total off-policyness (drift + sampler) rather than sampler mismatch alone. If you'd rather keep the semantics of the sync fields strictly separate, the alternative is new field names (e.g. `combined_importance_sampling_*`) at the cost of a second vocabulary. I lean toward reusing the sync names with a docstring stating the difference — same knobs, honestly documented — but this is the decision I don't want to make unilaterally.

Default: I'd ship it enabled with `sequence_mask` and `clip_max=3.0` to match sync GRPO's defaults, since AsyncGRPO is experimental and the current behavior (no correction) is the thing this issue flags.

If this direction works for you, I'll open a PR with the config fields, the gate, the segment-sum sequence mode, metrics, and CPU tests — including one that pins the motivating case: an unchanged policy under top-p sampling produces ρ ≈ S and the gate neutralizes it.

**Comment by mmjerge** [comment]
Draft PR implementing the proposal above: #7120. It takes the reuse-sync-names-with-honest-docs side of the naming fork; happy to rename the fields or adjust defaults per your preference, and I'll mark it ready once that's settled.
